### PR TITLE
Harden ReminderService response to Collection disappearing

### LINKED ISSUE #6666: ReminderService - Collection was null
https://couchdb.ankidroid.org/acralyzer/_design/acralyzer/index.html#/report-details/4d47e023-7d2e-454f-b6a4-a64aa0011d2a

2.12.beta5
```
java.lang.RuntimeException: Unable to start receiver com.ichi2.anki.services.ReminderService: java.lang.NullPointerException: Attempt to invoke virtual method 'com.ichi2.libanki.Decks com.ichi2.libanki.Collection.getDecks()' on a null object reference
at android.app.ActivityThread.handleReceiver(ActivityThread.java:4120)
at android.app.ActivityThread.access$1700(ActivityThread.java:272)
at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2066)
at android.os.Handler.dispatchMessage(Handler.java:107)
at android.os.Looper.loop(Looper.java:237)
at android.app.ActivityThread.main(ActivityThread.java:8016)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1076)
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'com.ichi2.libanki.Decks com.ichi2.libanki.Collection.getDecks()' on a null object reference
at com.ichi2.anki.services.ReminderService.onReceive(ReminderService.java:3)
at android.app.ActivityThread.handleReceiver(ActivityThread.java:4111)
... 8 more
```

### PR DESCRIPTION
**Pull request by mikehardy**
## Purpose / Description

The ReminderService was a bit fragile with regard to collections disappearing from underneath it, but for users that are set to synchronize collection at startup, this would happen quite a bit as the end of the sync closes the collection but the ReminderService may still be working.

So this PR probes the ReminderService to expose the poor null handling, then hardens it against that case

## Fixes
Fixes #6666

## Approach

Most of this PR is test code actually. The actual functional change is pretty small.

1. Re-factor our test suite in the area of null collection handling. Testing it was onerous before. It's easy now.
1. Add a test for ReminderService, and make sure a null collection is provided - this causes a crash
1. Fix up ReminderService, notifications are still present but now it no longer crashes

## How Has This Been Tested?

This is a race condition in essence so it is hard to test but with the above procedure I was able to create test conditions that show the problem every time.

## Learning

When you do Robolectric tests and you want to check the results of system activity (like "hey system, how many notifications do I have?" or "did that Toast post correctly?") You don't just new the Robolectric "shadow" objects that can answer your question.

You have to `shadowOf(SystemStateManagerObject);`  (e.g. "NotificationManager)).

Then you can say for instance `ShadowNotificationManager.getAllNotifications()` or whatever. If you just new the objects you are interrogating an empty object that is not really shadowing anything so you get the wrong answers. 

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/services/ReminderService.java [hunk 1]
```diff
@@ -70,10 +70,17 @@ public void onReceive(Context context, Intent intent) {
         // 0 is not a valid dconf id.
         final long dConfId = intent.getLongExtra(EXTRA_DECK_OPTION_ID, 0);
         if (dConfId == 0) {
+            Timber.w("onReceive - dConfId 0, returning");
+            return;
+        }
+
+        Collection col = CollectionHelper.getInstance().getCol(context);
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/services/ReminderService.java:77]: I believe getCol can throw due to permissions exceptions

File: AnkiDroid/src/main/java/com/ichi2/anki/services/ReminderService.java [hunk 2]
```diff
@@ -70,10 +70,17 @@ public void onReceive(Context context, Intent intent) {
...
+        if (null == col || !CollectionHelper.getInstance().colIsOpen()) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/services/ReminderService.java:78]: Nit: consolidate `.getInstance` to a variable

File: AnkiDroid/src/main/java/com/ichi2/anki/services/ReminderService.java [hunk 3]
```diff
@@ -114,7 +124,7 @@ public void onReceive(Context context, Intent intent) {
                         .setContentText(context.getResources().getQuantityString(
                                 R.plurals.reminder_text,
                                 total,
-                                CollectionHelper.getInstance().getCol(context).getDecks().name(deckId),
+                                deckDue.getFullDeckName(),
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/services/ReminderService.java:127]: Note: Don't think this matters as we only list the top level

File: AnkiDroid/src/main/java/com/ichi2/anki/services/ReminderService.java [hunk 4]
```diff
@@ -140,20 +151,19 @@ public static Intent getReviewDeckIntent(@NonNull Context context, long deckId)
...
     @Nullable
-    private List<AbstractSched.DeckDueTreeNode> getDeckOptionDue(Context context, long dConfId, boolean recur) {
+    private List<AbstractSched.DeckDueTreeNode> getDeckOptionDue(Collection col, long dConfId, boolean recur) {
 
-        Collection col = CollectionHelper.getInstance().getCol(context);
         // Avoid crashes if the deck option group is deleted while we
         // are working
-        if (col.getDecks().getConf(dConfId) == null) {
-            Timber.d("Deck option %s deleted while ReminderService was working. Ignoring", dConfId);
+        if (col.getDb() != null && col.getDecks().getConf(dConfId) == null) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/services/ReminderService.java:158]: If the database is null, then this case is not hit. Should this be an `== null OR`

File: AnkiDroid/src/main/java/com/ichi2/anki/services/ReminderService.java (hunk 2 above)
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/services/ReminderService.java:78]: This is the actual crash fix

File: AnkiDroid/src/main/java/com/ichi2/anki/services/ReminderService.java [hunk 5]
```diff
@@ -88,11 +95,13 @@ public void onReceive(Context context, Intent intent) {
 
         final NotificationManagerCompat notificationManager = NotificationManagerCompat.from(context);
         if (!notificationManager.areNotificationsEnabled()) {
+            Timber.v("onReceive - notifications disabled, returning");
             return;
         }
-        List<AbstractSched.DeckDueTreeNode> decksDue = getDeckOptionDue(context, dConfId, true);
+        List<AbstractSched.DeckDueTreeNode> decksDue = getDeckOptionDue(col, dConfId, true);
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/services/ReminderService.java:101]: this method has pokemon handling, nothing will come out except maybe a null return

File: AnkiDroid/src/main/java/com/ichi2/anki/services/ReminderService.java (hunk 3 above)
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/services/ReminderService.java:127]: turns out we can avoid accessing collection here entirely, the deck due node has the name

File: AnkiDroid/src/test/java/com/ichi2/anki/services/ReminderServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    public void testReminderServiceNothingDue() {
+        buildDefaultDeckReminders();
+        assertThat("No notifications exist", getNotificationManagerShadow().size() == 0);
+    }
+
+
+    @Test
+    public void testReminderServiceReviewsDue() {
+        addNoteUsingBasicModel("test front", "test back");
+        assertThat("No notifications exist", getNotificationManagerShadow().size() == 0);
```
Comment (david-allison-1) [inline AnkiDroid/src/test/java/com/ichi2/anki/services/ReminderServiceTest.java:33]: Just for fun:

I think this could be `isCollectionWithSize` or `isEmptyCollection` or `is(empty())`

File: AnkiDroid/src/main/java/com/ichi2/anki/services/ReminderService.java (hunk 4 above)
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/services/ReminderService.java:158]: I don't have the CollectionHelper here so I can't ask `isClosed`, however the db gets nulled on close, so this check should be equivalent and the col.getDecks should succeed

File: AnkiDroid/src/main/java/com/ichi2/anki/services/ReminderService.java (hunk 1 above)
Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/services/ReminderService.java:77]: I think might be right. Okay, I'll go "full hardening" then, not just the stuff I can easily demonstrate with the unit test I posted up

File: AnkiDroid/src/main/java/com/ichi2/anki/services/ReminderService.java (hunk 3 above)
Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/services/ReminderService.java:127]: While true, I wanted it to be NF just swapping in the member var access vs full-fetch-from-decks and this was I think the analogous name.

File: AnkiDroid/src/main/java/com/ichi2/anki/services/ReminderService.java (hunk 4 above)
Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/services/ReminderService.java:158]: Absolutely right, crucial error. Fixing

File: AnkiDroid/src/test/java/com/ichi2/anki/services/ReminderServiceTest.java (hunk 6 above)
Comment (mikehardy) [reply AnkiDroid/src/test/java/com/ichi2/anki/services/ReminderServiceTest.java:33]: true, but don't those expected notification counts line up so beautifully on the right? :star_struck:  - I'm going to leave it

File: AnkiDroid/src/main/java/com/ichi2/anki/services/ReminderService.java (hunk 1 above)
Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/anki/services/ReminderService.java:77]: We do have a `getColSafe`, but that sends an exception report. I worked around this previously in `BootService`

**Comment by mikehardy** [comment]
@david-allison-1 re-pushed with all comments addressed, thanks, esp for seeing the logic bug in the conditional! w.r.t. the test churn, luckily with alterations to existing tests you can easily do the "break first / verify test fails / small change to fix" loop so I'm pretty comfortable they are correct. Hopefully they are tasteful as well.

File: AnkiDroid/src/main/java/com/ichi2/anki/services/ReminderService.java (hunk 1 above)
Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/services/ReminderService.java]: For ReminderService - which I never ever ever ever want to block anything else from functioning - I like the idea of complete and total hardening so I just catch throwable here. ReminderService should just never crash, it's so ancillary
